flink-connectors/flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaWriter.java

```diff
@@ -406,6 +419,7 @@ public WriterCallback(
         @Override
         public void onCompletion(RecordMetadata metadata, Exception exception) {
             if (exception != null) {
+                numRecordsOutErrorsCounter.inc();
```

Comment:
I think it is not safe to call this here because it is called by the Kafka thread. The counter should only be updated by the mail box thread. In general, I am a bit skeptical about introducing this change because the maximum number of errors is 1 since the pipeline immediately fails after coming to this point.

Reply:
The metric is useful for user to be aware of the error even if it only have 1. I will move the call to the mail box thread.